subsys/logging/log_backend_fcb.c

```diff
@@ -0,0 +1,186 @@
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include "log_backend_std.h"
+#include <fs/fcb.h>
+#include <assert.h>
+
+static struct flash_sector sectors[CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT] = {
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR1_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR1_SIZE },
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 2
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR2_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR2_SIZE },
+#endif
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 3
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR3_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR3_SIZE },
+#endif
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 4
+        { .fs_off = CONFIG_LOG_BACKEND_FCB_SECTOR4_OFFSET,
+          .fs_size = CONFIG_LOG_BACKEND_FCB_SECTOR4_SIZE },
+#endif
+#if CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT >= 5
+#error "Max 4 log sectors are supported"
+#endif
+};
+
+static uint8_t buf[CONFIG_LOG_BACKEND_FCB_LINE_LEN];
+static struct fcb fcb = {
+        .f_magic = 0xfcb00106, /* "fcb log" */
+        .f_version = 1,
+        .f_sector_cnt = CONFIG_LOG_BACKEND_FCB_SECTOR_COUNT,
+        .f_sectors = sectors,
+};
+
+static int store_line(uint8_t *data, size_t length, void *ctx)
+{
+        struct fcb_entry loc;
+        int rc;
+
+        /* format: [00:00:04.388,000] <inf> this is a log line */
+        /* store only <wrn> and <err> lines. */
+        /* search for < to manage with and without colours */
+        char *p = memchr(data, '<', length);
+        if (!p || (p[1] != 'e' && p[1] != 'w'))
+                return length;
+
+        while (1) {
+                rc = fcb_append(&fcb, length, &loc);
+                if (!rc) {
+                        break;
+                }
+                else if (rc == -ENOSPC) {
+                        rc = fcb_rotate(&fcb);
+                        if (rc) {
+                                printk("fcb_rotate() returned %d\n", rc);
+                                return length;
+                        }
+                }
+                else {
+                        printk("fcb_append() returned %d\n", rc);
+                        return length;
+                }
+        }
+        rc = flash_area_write(fcb.fap, FCB_ENTRY_FA_DATA_OFF(loc),
+                              data, length);
+        if (rc) {
+                printk("flash_area_write() returned %d\n", rc);
+                return length;
+        }
+
+        rc = fcb_append_finish(&fcb, &loc);
+        if (rc) {
+                printk("fcb_append_finish() returned %d\n", rc);
+                return length;
+        }
+
+        return length;
+}
+
+LOG_OUTPUT_DEFINE(log_output, store_line, buf, sizeof(buf));
+
+static void put(const struct log_backend *const backend,
+		struct log_msg *msg)
+{
+        uint32_t flag = LOG_OUTPUT_FLAG_FORMAT_SYST;
+
+        log_backend_std_put(&log_output, flag, msg);
```

Comment:
have you considered storing just the message? it would use much less space. The only drawback is that duplicated strings would be lost because %s points to area in RAM. For first iteration it could be replaced with pointer to some constant string.

Printing/processing would also be easier since it would be done message by message.